// sync-server/lib/request.js
(function() {
  "use strict";
  var http = require('http')
    , req = http.IncomingMessage.prototype
    , parseUrl = require('url').parse;

  //Patch ServerRequest to save unmodified copy of headers
  var _addHeaderLine = req._addHeaderLine;
  req._addHeaderLine = function(field, value) {
    var list = this.complete ?
        (this.allTrailers || (this.allTrailers = {})) :
        (this.allHeaders || (this.allHeaders = {}));
    if (field in list) {
      list[field].push(value);
    } else {
      list[field] = [value];
    }
    _addHeaderLine.call(this, field, value);
  };

  //todo: move to qs module or global
  var urlDec = function(s) {
    s = String(s).replace(/\+/g, ' ');
    try {
      return decodeURIComponent(s);
    } catch(e) {
      return unescape(s);
    }
  };

  function Request(httpReq) {
    this._super = httpReq;
  }

  Request.prototype = {
    getURL: function() {
      return this._super.url;
    },
    getURLParts: function() {
      if (!this._parsed) {
        var parts = parseUrl(this._super.url);
        this._parsed = {path: urlDec(parts.pathname), qs: parts.search || ''};
      }
      return this._parsed;
    },
    getHeaders: function() {
      //var allHeaders = this._super.allHeaders || {};
      //var headers = parseHeaders(allHeaders);
      return this._super.headers;
    },
    getPostData: function() {
    },
    getCookies: function() {
    }
  };

  module.exports = Request;
})();
