// sync-server/lib/request.js
(function() {
  "use strict";
  var http = require('http')
    , req = http.IncomingMessage.prototype
    , _parseUrl = require('url').parse;

  //Patch ServerRequest to save unmodified copy of headers
  var _addHeaderLine = req._addHeaderLine;
  req._addHeaderLine = function(field, value) {
    var list = this.complete ?
        (this.allTrailers || (this.allTrailers = {})) :
        (this.allHeaders || (this.allHeaders = {}));
    if (field in list) {
      list[field].push(value);
    } else {
      list[field] = [value];
    }
    _addHeaderLine.call(this, field, value);
  };

  var parseUrl = function(url) {
    var parts = _parseUrl(url);
    //todo: urlDec()
    return {path: parts.pathname, qs: parts.search || ''};
  };

  var parseHeaders = function(headers) {
    //todo: deal with raw headers
    return headers;
  };

  function Request(httpReq) {
    this._super = httpReq;
  }

  Request.prototype = {
    getURL: function() {
      return this._super.url;
    },
    getURLParts: function() {
      return this._parsed || (this._parsed = parseUrl(this._super.url));
    },
    getHeaders: function() {
      return parseHeaders(this._super.allHeaders || {});
    },
    getPostData: function() {
    },
    getCookies: function() {
    }
  };

  module.exports = Request;
})();
